src/ImageSharp/Formats/Heif/HeifDecoderCore.cs

```diff
@@ -0,0 +1,684 @@
+                item?.DataLocations.Add(loc);
+            }
+        }
+    }
+
+    private static uint ReadUInt16Or32(Span<byte> buffer, bool isLarge, ref int bytesRead)
+    {
+        uint result;
+        if (isLarge)
+        {
+            result = BinaryPrimitives.ReadUInt32BigEndian(buffer[bytesRead..]);
+            bytesRead += 4;
+        }
+        else
+        {
+            result = BinaryPrimitives.ReadUInt16BigEndian(buffer[bytesRead..]);
+            bytesRead += 2;
+        }
+
+        return result;
+    }
+
+    private static ulong ReadUIntVariable(Span<byte> buffer, int numBytes, ref int bytesRead)
+    {
+        ulong result = 0UL;
+        if (numBytes == 8)
+        {
+            result = BinaryPrimitives.ReadUInt64BigEndian(buffer[bytesRead..]);
+            bytesRead += 8;
+        }
+        else if (numBytes == 4)
+        {
+            result = BinaryPrimitives.ReadUInt32BigEndian(buffer[bytesRead..]);
+            bytesRead += 4;
+        }
+        else if (numBytes == 2)
+        {
+            result = BinaryPrimitives.ReadUInt16BigEndian(buffer[bytesRead..]);
+            bytesRead += 2;
+        }
+        else if (numBytes == 1)
+        {
+            result = buffer[bytesRead++];
+        }
+
+        return result;
+    }
+
+    private Image<TPixel> ParseMediaData<TPixel>(BufferedReadStream stream, long boxLength)
+        where TPixel : unmanaged, IPixel<TPixel>
+    {
+        // FIXME: No HVC decoding yet, so parse only a JPEG thumbnail.
+        HeifItemLink? thumbLink = this.itemLinks.FirstOrDefault(link => link.Type == Heif4CharCode.Thmb);
+        if (thumbLink == null)
+        {
+            throw new NotImplementedException("No thumbnail found");
+        }
+
+        HeifItem? thumbItem = this.FindItemById(thumbLink.SourceId);
+        if (thumbItem == null || thumbItem.Type != Heif4CharCode.Jpeg)
+        {
+            throw new NotImplementedException("No HVC decoding implemented yet");
+        }
+
+        int thumbFileOffset = (int)thumbItem.DataLocations[0].Offset;
+        int thumbFileLength = (int)thumbItem.DataLocations[0].Length;
+        stream.Skip((int)(thumbFileOffset - stream.Position));
+        using IMemoryOwner<byte> thumbMemory = this.configuration.MemoryAllocator.Allocate<byte>(thumbFileLength);
+        Span<byte> thumbSpan = thumbMemory.GetSpan();
+        stream.Read(thumbSpan);
+
+        HeifMetadata meta = this.metadata.GetHeifMetadata();
+        meta.CompressionMethod = HeifCompressionMethod.LegacyJpeg;
+
+        return Image.Load<TPixel>(thumbSpan);
+    }
+
+    private static void SkipBox(BufferedReadStream stream, long boxLength)
+        => stream.Skip((int)boxLength);
+
+    private Span<byte> ReadIntoBuffer(BufferedReadStream stream, long length)
+    {
+        if (length <= this.buffer.Length)
+        {
+            stream.Read(this.buffer, 0, (int)length);
```

Comment:
`Stream.Read/ReadAsync(buffer)` is not guranteed to read `buffer.Length` bytes. Ignoring the return value may lead to bugs (#1707) or even worse, critical security vulnerabilities (#2550 fixed an infinite loop that could be triggered by premature EOF).

The return value should be **always** checked.

Reply:
Corrected all occurrences.